lua/nvim-treesitter/node_movement.lua

```diff
@@ -1,88 +1,76 @@
-local M = {}
 
+local configs = require'nvim-treesitter.configs'
+local state = require'nvim-treesitter.state'
+local ts_utils = require'nvim-treesitter.ts_utils'
+
+local M = {}
 
-M.NodeMovementKind = {
+local NodeMovementKind = {
   up = 'up',
   down = 'down',
   left = 'left',
   right = 'right',
 }
 
-M.current_node = {}
-
-local function node_start_to_vim(node)
-  if not node then return end
-
-  local row, col = node:start()
-  local exec_command = string.format('call cursor(%d, %d)', row+1, col+1)
-  api.nvim_exec(exec_command, false)
-end
+local get_node_fn = {
+  [NodeMovementKind.up] = function(node)
+    return node:parent()
+  end,
+  [NodeMovementKind.down] = function(node)
+    if node:named_child_count() > 0 then
+      return node:named_child(0)
+    else
+      local next_node = ts_utils.get_next_node(node)
+      if next_node and next_node:named_child_count() > 0 then
+        return next_node:named_child(0)
+      end
+    end
+  end,
+  [NodeMovementKind.left] = function(node)
+    return ts_utils.get_previous_node(node, true, true)
+  end,
+  [NodeMovementKind.right] = function(node)
+    return ts_utils.get_next_node(node, true, true)
+  end,
+}
 
 M.do_node_movement = function(kind)
-  local buf, line, col = unpack(vim.fn.getpos("."))
+  local buf = api.nvim_get_current_buf()
 
-  local current_node = M.current_node[buf]
+  local buf_state = state.get_buf_state(buf)
+  if not buf_state then return end
 
-  if current_node then
-    local node_line, node_col = current_node:start()
-    if line-1 ~= node_line or  col-1 ~= node_col then
-      current_node = nil
-    end
-  end
-  local destination_node
-
-  if parsers.has_parser() then
-    local root = parsers.get_parser():parse():root()
-    if not current_node then
-      current_node = root:named_descendant_for_range(line-1, col-1, line-1, col)
-    end
+  local current_node = buf_state.current_node
+  if not current_node then return end
 
-    if kind == M.NodeMovementKind.up then
-       destination_node = current_node:parent()
-    elseif kind == M.NodeMovementKind.down then
-      if current_node:named_child_count() > 0 then
-        destination_node = current_node:named_child(0)
-      else
-        local next_node = utils.get_next_node(current_node)
-        if next_node and next_node:named_child_count() > 0 then
-          destination_node = next_node:named_child(0)
-        end
-      end
-    elseif kind == M.NodeMovementKind.left then
-      destination_node = utils.get_previous_node(current_node, true, true)
-    elseif kind == M.NodeMovementKind.right then
-      destination_node = utils.get_next_node(current_node, true, true)
-    end
-    M.current_node[buf] = destination_node or current_node
-  end
+  local destination_node = get_node_fn[kind](current_node)
 
   if destination_node then
-    node_start_to_vim(destination_node)
+    local row, col = destination_node:start()
+    vim.fn.setpos(".", { buf, row+1, col+1, 0 })
```

Comment:
Could you set `m'`? Apparently, this is the offical ways to set jumps though it was not working in every case for me

Reply:
in the documentation they explicitly say setpos doesn't allow setting marks. i'll try it though.